R-package/R/lgb.train.R

```diff
@@ -124,6 +124,60 @@ lgb.train <- function(params = list(),
     end_iteration <- begin_iteration + nrounds - 1L
   }
 
+  # Convert interaction constraints to feature numbers
+  if (!is.null(params[["interaction_constraints"]])) {
```

Comment:
thanks for this! 

@btrotta I'll leave some comments here and you can decide, but I think it would make sense to do this in a separate pull request so it doesn't hold this one up.

* Unfortunately we have a lot of duplication between `lgb.train()` and  `lgb.cv()` today 
 (https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.cv.R), so handling for this parameter needs to be added in both places. I think it would make sense to create a new internal function like  `.check_interaction_constraints()`, similar to how we have [`lgb.check.obj()`](https://github.com/microsoft/LightGBM/blob/5304885c0aa4f26e29ae2a4d451c0fb39d70f24b/R-package/R/utils.R#L186), that is then called by both functions. That would also make it easier to unit test the handling of this parameter.
* We should add some validation that `interaction_constraints` is a list of either numeric or character vectors. Like this

```r
interaction_constraints <- params[["interaction_constraints"]]
if (!is.null(interaction_constraints)){
    # validation
    if (!methods::is(interaction_constraints, "list")){
        stop("interaction_constraints must be a list")
    }
    if (!all(sapply(interaction_constraints, function(x){is.character(x) || is.numeric(x)}))){
        stop("every element in interaction_constraints must be a character vector or numeric vector)
    }

    # rest of the code
}
```

*  I think the uses of `as.list()` in this code should be replaced with `as.integer()`
* I'd like too have unit tests, for both `lgb.train()` and `lgb.cv()`. That will catch some issues and prevent us from breaking this functionality in the future.
    - validation that you get an informative error if you don't pass a `list` object
    - validation that you get an informative error if you pass a `list` of something other than character vectors and numeric vectors
    - validation that you get an informative error for indices that are greater than `max_features`
    - test training with only passing indices
    - test training with only passing feature names
    - test training with passing a mix of indices and feature names

Reply:
Ok sure, I have moved these changes to new PR #3136.